File: 21553 - Sai Nandan D/Week 6/2416_Sum of Prefix Scores of Strings.py
class TrieNode:
    
    def __init__(self):
        self.children = collections.defaultdict(TrieNode)
        self.count = 0
            
class Trie:

    def __init__(self):
        self.root = TrieNode()

    def insert(self, word: str) -> None:
        node = self.root
        for letter in word:
            node = node.children[letter]
            node.count += 1            

    def score(self, word: str) -> bool:
        node = self.root
        total = 0
        for letter in word:            
            node = node.children[letter]
            total += node.count
        return total
    

class Solution:                                        
    def sumPrefixScores(self, words: List[str]) -> List[int]:
        t = Trie()
        for w in words:
            t.insert(w)
            
        return [t.score(w) for w in words]

File: 21551-naveen-siva-kumar/week6/2416-sum-of-prefix-scores-of-strings/2416-sum-of-prefix-scores-of-strings.cpp

class TrieNode{
    public:
    unordered_map<char, TrieNode*> children;
    char val;
    bool isEnd = false;
    int count =0;
    
    TrieNode(){}
    
    TrieNode(char v){
        this->val = v;
    }
};
   
class Trie {
    
    public:
    TrieNode* root;

    
    
    public:
        Trie() {
        root = new TrieNode();
        }

        void insert(string word) {
           TrieNode* node = root;
            for(int i = 0; i < word.size(); i++){
                char c = word[i];
                if(node->children.find(c) == node->children.end()){
                    node->children[c] = new TrieNode(c);                        
                }

                node = node->children[c]; 
                node->count++;
            }

            node->isEnd = true;   
        }

        bool search(string word) {
            TrieNode* node = root;
            for(int i = 0; i < word.size(); i++){
                char c = word[i];
                if(node->children.find(c) == node->children.end()) return false;
                node = node->children[c];
            }

            return node->isEnd; 
        }

        bool startsWith(string prefix) {
            TrieNode* node = root;
            for(int i = 0; i < prefix.size(); i++){
                char c = prefix[i];
                if(node->children.find(c) == node->children.end()) return false;
                node = node->children[c];
            }

            return true;
        }
        int count(string word){
            int count =0;
            TrieNode* node = root;
            for(int i = 0; i < word.size(); i++){
                char c = word[i];
                //if(node->children.find(c) == node->children.end()) return false;
                node = node->children[c];
                count +=node->count;
            }

            return count; 
        }
};





class Solution {
    
    
public:
    vector<int> sumPrefixScores(vector<string>& words) {
        
        Trie* root = new Trie();
        
        for(int i=0;i<words.size();i++){
            root->insert(words[i]);
        }
        vector<int> ans;
        int count =0;
        for(int i=0;i<words.size();i++){
            
            ans.push_back(root->count(words[i]));
            
        }
        return ans;
    }
};